File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D06.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.math.quadraticFormula
import org.gristle.pdxpuzzles.utilities.parsing.getLongList
import kotlin.math.ceil
import kotlin.math.floor

class Y23D6(private val input: String) : Day {

    override fun part1() = solve { it.getLongList() }

    override fun part2() = solve { listOf(it.filter(Char::isDigit).toLong()) }

    private inline fun solve(parseLine: (String) -> List<Long>): Long {
        
        // parses input by parsing the digits in each input line according to the rules for each part, then
        // transposing the list of lists so that the vertical columns are grouped together.
        val races: List<Pair<Long, Long>> = input
            .lines()
            .map(parseLine)
            .let { (times, distances) -> times zip distances }
        
        return races.map(::waysToWin).reduce(Long::times)
    }

    private fun waysToWin(race: Pair<Long, Long>): Long {
        val (time, distance) = race
        
        // equation can be expressed as quadratic equation
        // D = distance, T = time, t = time pushing button
        // D = t(T - t) => (-1)t^2 + (T)t + (-D) = 0
        val (root1, root2) = quadraticFormula(-1, time, -distance)
        return (ceil(root2 - 1.0) - floor(root1 + 1.0)).toLong() + 1L
    }
}

fun main() = Day.runDay(Y23D6::class)

//    Class creation: 2ms
//    Part 1: 2374848 (2ms)
//    Part 2: 39132886 (0ms)
//    Total time: 5ms

@Suppress("unused")
private val sampleInput = listOf(
    """Time:      7  15   30
Distance:  9  40  200
""" to "288",
)

File: rust/advent/src/bin/2023/y23d06.rs
use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = i64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 6).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve<F>(input: &str, parse_line: F) -> i64
    where
        F: Fn(&str) -> Vec<i64>
{
    let races: Vec<Vec<i64>> = input
        .lines()
        .map(parse_line)
        .collect();

    let times = races.get(0).unwrap();
    let distances = races.get(1).unwrap();
    let races = times.iter().zip(distances.iter());
    races.map(|(time, distance)| ways_to_win(*time, *distance))
        .product()
}

fn ways_to_win(time: i64, distance: i64) -> i64 {
    let b = time as f64;
    let c = -distance as f64;
    let (root_1, root_2) = quadratic(-1.0, b, c);
    ((root_2 - 1.0).ceil() - (root_1 + 1.0).floor()) as i64 + 1
}

fn quadratic(a: f64, b: f64, c: f64) -> (f64, f64) {
    let determinant = (b * b - 4.0 * a * c).sqrt();
    let root_1 = (-b + determinant) / (2.0 * a);
    let root_2 = (-b - determinant) / (2.0 * a);
    (root_1, root_2)
}


fn part1(input: Input) -> Output {
    let parse_line = |line: &str| -> Vec<i64> {
        line.split_whitespace().filter_map(|substr| substr.parse().ok()).collect()
    };
    solve(input, parse_line)
}

fn part2(input: Input) -> Output {
    let parse_line = |line: &str| -> Vec<i64> {
        let digits: String = line.chars().filter(|c| c.is_ascii_digit()).collect();
        let number = digits.parse::<i64>().unwrap();
        vec![number]
    };
    solve(input, parse_line)
}

#[test]
fn default() {
    let input = get_input(23, 6).unwrap();
    assert_eq!(2374848, part1(&input));
    assert_eq!(39132886, part2(&input));
}
